### PR TITLE
Increase details of Channels documentation

### PR DESCRIPTION
**Pull request by oxzi**
With a release approaching, the Channels documentation got more information about how the Channel system actually work and how to build your own channels.

The internal/changes/examples JSON files were inlined to ease the readability and updated, as they contained partially outdated information.

For the pkg/plugin part, some API documentation was added as this might get used by external users.

### CODE REVIEW

File: doc/10-Channels.md [hunk 1]
```diff
@@ -1,88 +1,322 @@
...
+For logging or debugging, channels can write to `stderr`, which will be logged by Icinga Notifications.
 
 ### Request
 
-The request must be in JSON format and should contain following keys:
+A channel receives a request as a JSON object with the following fields:
 
-- `method`: The request method to call.
-- `params`: The params for request method.
+- `method`: The request method to call.
```
[the remark this answers is not in the document]
Comment (oxzi) [reply doc/10-Channels.md:35]: :<

File: doc/10-Channels.md [hunk 2]
```diff
@@ -1,88 +1,322 @@
...
+          "en_US": "Response Status Codes"
+        },
+        "help": {
+          "de_DE": "Kommaseparierte Liste erwarteter Status-Code der HTTP-Antwort, z.B.: 200,201,202,208,418",
+          "en_US": "Comma separated list of expected HTTP response status code, e.g., 200,201,202,208,418"
+        },
+        "default": "200",
+        "min": null,
+        "max": null
+      }
```
[the remark this answers is not in the document]
Comment (oxzi) [reply doc/10-Channels.md:187]: Valid point. It was the output from the webhook channel. I have now reduced it to two options, being used later on.

File: doc/10-Channels.md [hunk 3]
```diff
@@ -1,88 +1,298 @@
...
+
+Each request must be answered by the channel with a response JSON object of the following fields:
+
+- `result`: The result as JSON format. Omitted when the method does not return a value, i.e., for setter calls, or an
+  error has occurred.
+- `error`: The error message. Omitted when no error has occurred.
+- `id`: The request id.
+
+In case of a present `error` value, the `result` field should be empty.
+Successfully responses without a `result` do only contain the `id` field.
```
Comment (yhabteab) [inline doc/10-Channels.md:70]: This (few other places as well) should be rephrased as [`Successful`](https://dictionary.cambridge.org/dictionary/english/successful) instead.

File: doc/10-Channels.md [hunk 4]
```diff
@@ -1,88 +1,298 @@
...
+#### GetInfo
+
+The parameterless `GetInfo` method returns information about the channel.
+
+Its `result` is expected to be a JSON object with the `json` fields defined in the
+[`Info` type](https://pkg.go.dev/github.com/icinga/icinga-notifications/pkg/plugin#Info).
+The `config_attrs` field must be an array of JSON objects according to the
+[`ConfigOption` type](https://pkg.go.dev/github.com/icinga/icinga-notifications/pkg/plugin#ConfigOption).
+Those attributes define configuration options for the channel to be set via the `SetConfig` method.
+Furthermore, they are used to build the channel configuration UI in Icinga Notifications Web.
```
Comment (yhabteab) [inline doc/10-Channels.md:108]: I think that `UI` is quite out of place. You might wanted to rephrase it as `... build the UI channel configuration in ...`?

File: pkg/plugin/plugin.go [hunk 5]
```diff
@@ -69,57 +69,74 @@ type ConfigOption struct {
...
 	Addresses []*Address `json:"addresses"`
 }
 
+// Address to receive this notification. Each Contact might have multiple addresses.
 type Address struct {
 	Type    string `json:"type"`
 	Address string `json:"address"`
 }
 
+// Object which this NotificationRequest is all about, e.g., a service check.
```
Comment (yhabteab) [inline pkg/plugin/plugin.go:105]: I don't think a `service check` is an object.

Comment (oxzi) [reply pkg/plugin/plugin.go:105]: The text says that the object might be _about_ a service check, not being a service check.

File: doc/10-Channels.md (hunk 4 above)
Comment (oxzi) [reply doc/10-Channels.md]: I have rephrased the sentence.

File: pkg/plugin/plugin.go (hunk 5 above)
Comment (yhabteab) [reply pkg/plugin/plugin.go:105]: > The text says that the object might be about a service check

That sentence doesn't make sense to me either. An `Event` could be about a service check, but I don't quite understand what an object being about a service check is supposed to mean.

Comment (oxzi) [reply pkg/plugin/plugin.go:105]: Then please take a look at the example `object`: <https://github.com/Icinga/icinga-notifications/blob/17cb89f00b9618c118bc91e2689dd98c65567980/doc/10-Channels.md#example-sendnotification-request>.

I am also open for alternative suggestions.

Comment (oxzi) [reply pkg/plugin/plugin.go:105]: Or was your point that it does not refer to a service check but more to a service object?

_Btw, I find the name `object` hard to grasp, especially as both in Icinga 2 and OOP lingo everything is a object._

File: doc/10-Channels.md [hunk 6]
```diff
@@ -1,88 +1,298 @@
...
 
 ### Request
 
-The request must be in JSON format and should contain following keys:
+A channel receives a request as a JSON object with the following fields:
 
 - `method`: The request method to call.
-- `params`: The params for request method.
+- `params`: Optional params for the request method.
 - `id`: Unsigned int value. Required to assign the response to its request as responses can be sent out of order.
```
Comment (martialblog) [inline doc/10-Channels.md:37]: Hi, 

the field description could be normalized:

`fieldname: description of what it does. optional/madatory. data type`

Regards

File: doc/10-Channels.md [hunk 7]
```diff
@@ -1,88 +1,298 @@
...
+    },
+    "id": 1000
+  }
+  ```
+
+### Response
+
+Each request must be answered by the channel with a response JSON object of the following fields:
+
+- `result`: The result as JSON format. Omitted when the method does not return a value, i.e., for setter calls, or an
```
Comment (martialblog) [inline doc/10-Channels.md:64]: Again, normalization could be useful:

`fieldname: description of what it does. optional/madatory. data type`

File: doc/10-Channels.md [hunk 8]
```diff
@@ -1,88 +1,298 @@
...
+  error has occurred.
+- `error`: The error message. Omitted when no error has occurred.
+- `id`: The request id.
+
+In case of a present `error` value, the `result` field should be empty.
```
Comment (martialblog) [inline doc/10-Channels.md:69]: Does "empty" mean `null` or the field can be omitted. Or does it mean empty string?

File: doc/10-Channels.md [hunk 9]
```diff
@@ -1,88 +1,298 @@
...
   },
-  "id": 2020
+  "id": 1
 }
 ```
 
+#### SetConfig
+
+The `SetConfig` method configures the channel and
+should be called at least once before sending the first notifications to initialize the channel plugin.
```
Comment (martialblog) [inline doc/10-Channels.md:167]: "should be called" by whom? The channel plugin?

File: doc/10-Channels.md [hunk 10]
```diff
@@ -1,88 +1,298 @@
...
+To illustrate, the URL template from the above output is configurable with the following JSON object passed in `params`:
+
 ```json
 {
-  "method": "Foo",
+  "url_template": "http://localhost:8000/update/{{.Incident.Id}}"
+}
+```
+
+The response might carry an error if the configuration is invalid.
```
Comment (martialblog) [inline doc/10-Channels.md:181]: What does "might" mean here? Optional? Recommended?

Might I suggest using the key words from RFC2119?

File: doc/10-Channels.md [hunk 11]
```diff
@@ -1,88 +1,298 @@
...
+#### SendNotification
 
-Examples:
+The `SendNotification` method requests the channel to dispatch notifications.
+
+Within the request's `params`, a JSON object representing a
+[`NotificationRequest`](https://pkg.go.dev/github.com/icinga/icinga-notifications/pkg/plugin#NotificationRequest)
+is expected.
+
+The response might carry an error due to channel-specific reasons.
```
Comment (martialblog) [inline doc/10-Channels.md:212]: Maybe add an example of what an error response

File: doc/10-Channels.md (hunk 6 above)
Comment (yhabteab) [reply doc/10-Channels.md:37]: Yes, maybe as a markdown table, like [Icinga 2](https://icinga.com/docs/icinga-2/latest/doc/12-icinga2-api/#creating-config-objects) does.

Comment (oxzi) [reply doc/10-Channels.md:37]: Thanks., I liked this idea. I tried to unify all fields to:
> name: Short description. Type.

File: doc/10-Channels.md (hunk 9 above)
Comment (oxzi) [reply doc/10-Channels.md:167]: Good point. In the eyes of a channel developer, this is straight out confusing.

File: doc/10-Channels.md (hunk 10 above)
Comment (oxzi) [reply doc/10-Channels.md:181]: `s/might/MAY/g` :^)

I rewrote some sections to remove inaccuracies, use some RFC 2119 language (just w/o the caps), and to address a channel developer.

File: doc/10-Channels.md (hunk 6 above)
Comment (oxzi) [reply doc/10-Channels.md]: I switched to a table, as you suggested @yhabteab. This has the only downside that diffs altering the column width span over the whole table. But, it looks nicer :)

Comment (oxzi) [reply doc/10-Channels.md]: @yhabteab: I tried to mimic Icinga 2's layout.

**Comment by oxzi** [comment]
I reworked some parts, especially to mention changes from #206. Furthermore, I described how the channel configuration through the two RPC methods works and how a channel plugin is expected to behave. As the documentation links to the Go API doc, I added more API docs for the fields.

Please take another look.

File: doc/10-Channels.md [hunk 12]
```diff
@@ -1,88 +1,355 @@
...
+
+```json
+{
+  "url_template": "http://localhost:8000/update/{{.Incident.Id}}"
 }
 ```
 
+If the channel plugin successfully configured itself, a response without an `result` must be returned.
+Otherwise, if the channel decides that the provided configuration is faulty, an `error` must be returned.
+This may happen if, i.e., an invalid configuration value was given.
```
Comment (yhabteab) [inline doc/10-Channels.md:200]: I think you don't want to use [`i.e.`](https://dictionary.cambridge.org/dictionary/english/ie) here, but `for instance/example` instead.

File: doc/10-Channels.md [hunk 13]
```diff
@@ -1,88 +1,317 @@
...
+    "id": 1000
+  }
+  ```
 
-- `method`: The request method to call.
-- `params`: The params for request method.
-- `id`: Unsigned int value. Required to assign the response to its request as responses can be sent out of order.
+### Response
+
+Each request must be answered by the channel with a response JSON object of the following fields:
```
[the remark this answers is not in the document]
Comment (oxzi) [reply doc/10-Channels.md:74]: I removed the "response" but kept JSON object to exactly specify what we are talking about. I don't want to rely on too much implied information.

File: doc/10-Channels.md (hunk 12 above)
Comment (oxzi) [reply doc/10-Channels.md:200]: There will be a day when I finally learn how to correctly use _e.g._ and _i.e._

**Review by yhabteab — APPROVED** [review]
LFTM!

**Review by julianbrost — APPROVED** [review]

**Comment by julianbrost** [comment]
@martialblog Do you want to have another look as well, i.e. should I wait with merging this a bit?

**Comment by martialblog** [comment]
@julianbrost Thanks, I was away these last days. If I find something I'll just open a PR. 

Looking forward to v0.1.0

**Pull request merged**